### PR TITLE
Fix date for year < 1980 and > 2108

### PR DESCRIPTION
**Pull request by mat-c**
Change-Id: Ifcc1be945bda1cd0e0dc7ae71a26c939135e51f0

Signed-off-by: Matthieu Castet <castet.matthieu@free.fr>

### CODE REVIEW

**Comment by mat-c** [comment]
Without this patch, I get "Jan  1  2098" date on system with start at Jan 1 1970.

File: misc.c [hunk 1]
```diff
@@ -131,8 +131,16 @@ void exfat_set_entry_time(struct exfat_sb_info *sbi, struct timespec *ts,
...
+	if (tm.tm_year < 80) {
+		t = 0;
+		d = (0 <<  9) | (1 << 5) | 1;
+	} else if (tm.tm_year >= 208) {
+		t = (23 << 11) | (59 << 5) | (59 >> 1);
+		d = (127 <<  9) | (12 << 5) | 31;
+	} else {
+		t = (tm.tm_hour << 11) | (tm.tm_min << 5) | (tm.tm_sec >> 1);
+		d = ((tm.tm_year - 80) <<  9) | ((tm.tm_mon + 1) << 5) | tm.tm_mday;
+	}
```
Comment (namjaejeon) [inline misc.c:143]: This code is needed when kernel version is lower than 5.4 version ?

Looks this code.
#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0)
        sb->s_time_gran = 10 * NSEC_PER_MSEC;
        sb->s_time_min = EXFAT_MIN_TIMESTAMP_SECS;
        sb->s_time_max = EXFAT_MAX_TIMESTAMP_SECS;
#endif

**Comment by mat-c** [comment]
Yes it is work 4.9 kernel. I didn't see the 5.4 part. I updated the patch

**Comment by namjaejeon** [comment]
@mat-c You have ever got the below build error before ?


In file included from fs/exfat/misc.c:14:
fs/exfat/misc.c: In function 'exfat_set_entry_time':
fs/exfat/exfat_raw.h:166:37: error: overflow in conversion from 'long long int' to '__kernel_time_t' {aka 'long int'} changes value from '4354819199' to '59851903' [-Werror=overflow]
  166 | #define EXFAT_MAX_TIMESTAMP_SECS    4354819199LL
      |                                     ^~~~~~~~~~~~
fs/exfat/misc.c:135:16: note: in expansion of macro 'EXFAT_MAX_TIMESTAMP_SECS'
  135 |   ts->tv_sec = EXFAT_MAX_TIMESTAMP_SECS;
      |                ^~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
scripts/Makefile.build:258: recipe for target 'fs/exfat/misc.o' failed

**Comment by namjaejeon** [comment]
@mat-c This issue seems to be coming from under 5.4 kernel version.

**Comment by mat-c** [comment]
Hello,

on which architecture did that happen ?

__kernel_time_t seems to be 32 bits.

Regards,

Matthieu

----- Mail original -----
De: "Namjae Jeon" ***@***.***>
À: "namjaejeon/linux-exfat-oot" ***@***.***>
Cc: "mat-c" ***@***.***>, "Mention" ***@***.***>
Envoyé: Mercredi 2 Juin 2021 10:05:41
Objet: Re: [namjaejeon/linux-exfat-oot] Fix date for year < 1980 and > 2108 (#17)





@mat-c You have ever got the below build error before ? 

In file included from fs/exfat/misc.c:14: 
fs/exfat/misc.c: In function 'exfat_set_entry_time': 
fs/exfat/exfat_raw.h:166:37: error: overflow in conversion from 'long long int' to '__kernel_time_t' {aka 'long int'} changes value from '4354819199' to '59851903' [-Werror=overflow] 
166 | #define EXFAT_MAX_TIMESTAMP_SECS 4354819199LL 
| ^~~~~~~~~~~~ 
fs/exfat/misc.c:135:16: note: in expansion of macro 'EXFAT_MAX_TIMESTAMP_SECS' 
135 | ts->tv_sec = EXFAT_MAX_TIMESTAMP_SECS; 
| ^~~~~~~~~~~~~~~~~~~~~~~~ 
cc1: all warnings being treated as errors 
scripts/Makefile.build:258: recipe for target 'fs/exfat/misc.o' failed

**Comment by namjaejeon** [comment]
ARM 32bit and linux 4.1 kernel.

**Comment by mat-c** [comment]
#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0)
void exfat_set_entry_time(struct exfat_sb_info *sbi, struct timespec64 *ts,
        u8 *tz, __le16 *time, __le16 *date, u8 *time_cs)
#else
void exfat_set_entry_time(struct exfat_sb_info *sbi, struct timespec *ts,
        u8 *tz, __le16 *time, __le16 *date, u8 *time_cs)
#endif



For kernel less than 4.19, we use struct timespec, and tv_sec is 32 bits on some platform (long int) !

For these platform we should clamp EXFAT_MAX_TIMESTAMP_SECS to 32 bits.



----- Mail original -----
De: "Namjae Jeon" ***@***.***>
À: "namjaejeon/linux-exfat-oot" ***@***.***>
Cc: "mat-c" ***@***.***>, "Mention" ***@***.***>
Envoyé: Jeudi 3 Juin 2021 03:57:03
Objet: Re: [namjaejeon/linux-exfat-oot] Fix date for year < 1980 and > 2108 (#17)





ARM 32bit and linux 4.1 kernel.

**Comment by namjaejeon** [comment]
Okay, Could you please provide the patch to fix it ? This issue is come from your patch.

**Pull request merged**